### PR TITLE
feat(core): per-piece exercise context derivation (#1081 B1)

### PR DESCRIPTION
**Pull request by jonyardley**
## What

B1 of the "lesson to mastery" epic (#1087, workstream **B** #1081): derive, per exercise, the piece and standalone contexts it has actually been practised in — the data layer B2 renders as "By piece" rows.

For each exercise entry in past sessions, the context is the **piece sharing its block `group_id`** in that same session (related-first, piece-last — resolved order-independently), or the **"On its own"** bucket when the entry was ungrouped or its block carried no piece. Surfaces as:

- `LibraryItemView.exercise_contexts: Vec<ExerciseContextView>` (populated for exercises; empty for pieces)
- `ExerciseContextView { piece: Option<PieceRefView>, latest_score, session_count, last_practiced_at }`
- Rollup v1 = latest recorded score per context; distinct-session count; most-recent date. Deterministic order: piece contexts (most-recent first, tie-broken by id), then "On its own" last.

Core-only, no UI.

## Invariants

- **6 (both modes):** pure ViewModel projection over `model.sessions` — identical whether sessions arrive online (`SessionsLoaded`) or local-first (`SessionsStoreLoaded`). Tested both paths.
- **Codec guard (#846 class):** `group_id` is the sole derivation input; a silent drop on the FFI wire erases every piece context. Added bincode round-trip guards for `SetlistEntry` with `group_id` set, and for `ExerciseContextView`.

## Tests (TDD — red first, then green)

- `test_exercise_contexts_derive_piece_and_on_its_own` — piece rollup (latest score across sessions, count, date) + "On its own" bucket + ordering + live composer subtitle + pieces carry no contexts.
- `test_exercise_contexts_grouped_without_piece_is_on_its_own` — a piece-less block falls to the bucket, not a phantom piece.
- `test_exercise_contexts_identical_in_both_modes` — invariant 6, online vs local-first load paths.
- Two FFI bincode round-trip guards.

## Coverage

Full — new derivation has both-modes + edge-case tests; new bridge types have round-trip guards. `build_view` wiring exercised via `app.view`.

## Deferred (tracked)

- #1093 — piece identity: snapshot title vs live composer, and tombstoned-piece filtering — decide during B2.
- #1094 — cache `exercise_contexts` on session-change instead of per-render (perf; negligible at current scale).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/intrada-core/src/domain/session.rs [hunk 1]
```diff
@@ -5034,4 +5034,32 @@
...
     }
+
+    /// `group_id` rides `SetlistEntry` across the bincode FFI wire (in the
+    /// persisted `PracticeSession` and the ViewModel). It is the sole input to
+    /// the B1 context derivation, so a silent drop on the wire would erase every
+    /// piece context. Guard the whole entry, with `group_id` set, against the
+    /// #846 class.
+    #[test]
+    fn setlist_entry_with_group_id_round_trips_on_ffi_bincode_wire() {
+        crate::domain::types::assert_round_trips(SetlistEntry {
```
[the remark this answers is not in the document]
Comment (jonyardley) [reply crates/intrada-core/src/domain/session.rs:5045]: Fixed. Root cause: the branch was cut from the older main where `assert_round_trips` was a shared `pub(crate)` module-level helper; the #1092 revert (now on main) moved it inside `mod tests` as private since it had no cross-module callers left. Rebased onto current main and hoisted it back to a `#[cfg(test)] pub(crate)` module-level helper — B1 legitimately re-introduces cross-module callers (session.rs + model.rs round-trip guards).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: crates/intrada-core/src/model.rs [hunk 2]
```diff
@@ -602,6 +626,22 @@
     use super::*;
     use chrono::Utc;
 
+    /// `ExerciseContextView` crosses the bincode FFI bridge inside the
+    /// ViewModel; guard it against the #846 silent-drop class.
+    #[test]
+    fn exercise_context_view_round_trips_on_ffi_bincode_wire() {
+        crate::domain::types::assert_round_trips(ExerciseContextView {
```
[the remark this answers is not in the document]
Comment (jonyardley) [reply crates/intrada-core/src/model.rs:633]: Fixed. Root cause: the branch was cut from the older main where `assert_round_trips` was a shared `pub(crate)` module-level helper; the #1092 revert (now on main) moved it inside `mod tests` as private since it had no cross-module callers left. Rebased onto current main and hoisted it back to a `#[cfg(test)] pub(crate)` module-level helper — B1 legitimately re-introduces cross-module callers (session.rs + model.rs round-trip guards).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
